File: scala/gigasecond/src/main/scala/Gigasecond.scala
import java.util.GregorianCalendar
import Gigasecond._

case class Gigasecond(start: Date) {
  lazy val date: Date = addSeconds(start, OneGigaSec)
}

object Gigasecond {
  type Date = GregorianCalendar
  type Seconds = Long
  type Milliseconds = Long

  val OneGigaSec: Seconds = math.pow(10, 9) toLong

  val MillisPerSecond = 1000L

  def addSeconds(date: Date, seconds: Seconds): Date = {
    val dateInMilliseconds: Milliseconds = date.getTime.getTime
    val timestamp: Milliseconds = dateInMilliseconds + seconds * MillisPerSecond

    val result = date.clone.asInstanceOf[Date]
    result.setTimeInMillis(timestamp)
    result
  }
}

File: java/gigasecond/src/main/java/Gigasecond.java
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

public class Gigasecond {

    private LocalDateTime startDateTime;

    public Gigasecond(LocalDate startDate) {
        this.startDateTime = LocalDateTime.of(startDate, LocalTime.of(0, 0));
    }

    public Gigasecond(LocalDateTime startDateTime) {
        this.startDateTime = startDateTime;
    }

    private static final long GIGA_SECONDS = (long)Math.pow(10, 9);

    public LocalDateTime getDate() {
        return startDateTime.plusSeconds(GIGA_SECONDS);
    }
}
